{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4fc14471-fcb6-461c-9d23-c744c6faee90",
   "metadata": {},
   "source": [
    "Here are customized functions to detect transmission events on a subset of samples on which you run StrainPhlAn"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ef7fc4a-cf68-4ac6-a12f-db58e61d2e76",
   "metadata": {},
   "source": [
    "**This is only compatible with MetaPhlAn (StrainPhlAn) 4.0.6** \\\n",
    "**Make sure you replace \"[YOUR CONDA ENVIRONMENT FOLDER]\" in your function with the path of your conda environment with MetaPhlAn 4.0.6 installed**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45e1e849-5c09-4de3-a73e-6619a18aa14c",
   "metadata": {},
   "source": [
    "How it works:\n",
    "For example, your goal is to detect transmission events for clade SGB1877 on a subset of your samples, make sure you have run StrainPhlAn for SGB1877\n",
    "1. First you run `extract_seq` function to extract the subset samples of interest from \"[Your StrainPhlAn output directory]/SGB1877/SGB1877.StrainPhlAn4_concatenated.aln\", the filtered fasta will be \"[Your StrainPhlAn output directory]/SGB1877/SGB1877.filtered.fasta\"\n",
    "2. Then, you can run `run_raxml` to make phylogenetic inference on the filtered fasta\n",
    "3. Before running the transmission detection function, replace the 'strain_transmission.py' in the conda environment with the customized 'strain_transmission.py' file. This is to set 0.1 or (0.01, 0.05) threshold for clades that were not in VALLES 2023 PAPER [PMID: 36653448]\n",
    "4. Run the `detect_transmission` function to detect transmission event "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6c93c259-4e81-413e-a67a-7968016b98e3",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'Bio'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mBio\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m SeqIO\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01msubprocess\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mos\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'Bio'"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "import subprocess\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6178cb48-1f91-4f39-9110-a2ed9aaecd6d",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'os' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mextract_seq\u001b[39m(strain_clade, sample_list,\n\u001b[0;32m----> 2\u001b[0m                 strainphlan_dir\u001b[38;5;241m=\u001b[39m\u001b[43mos\u001b[49m\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mabspath(strain_raw_dir),\n\u001b[1;32m      3\u001b[0m                 output_dir\u001b[38;5;241m=\u001b[39mos\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mabspath(strain_ana)):\n\u001b[1;32m      4\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;124;03m    Extract sequences of input clade from strainphlan alignment\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;124;03m    --------------------------------------------------\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;124;03m    \u001b[39;00m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m     18\u001b[0m     \u001b[38;5;66;03m# set up the path for the strainphlan alignment of the clade\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'os' is not defined"
     ]
    }
   ],
   "source": [
    "def extract_seq(strain_clade, sample_list,\n",
    "                strainphlan_dir=os.path.abspath(strain_raw_dir),\n",
    "                output_dir=os.path.abspath(strain_ana)):\n",
    "    \"\"\"\n",
    "    Extract sequences of input clade from strainphlan alignment\n",
    "    --------------------------------------------------\n",
    "    strain_clade: str\n",
    "        strain clade name\n",
    "    sample_list: list\n",
    "        list of samples included\n",
    "    strainphlan_dir: str\n",
    "        strainphlan raw data folder\n",
    "    output_dir: str\n",
    "        output folder to store the clade folders for the filtered fasta\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    # set up the path for the strainphlan alignment of the clade\n",
    "    strainphlan_fasta_path = f'{strainphlan_dir}/{strain_clade}/{strain_clade}.StrainPhlAn4_concatenated.aln'\n",
    "\n",
    "    # create the folder for output clade folder of the filtered fasta\n",
    "    output_clade_dir = f'{output_dir}/{strain_clade}'\n",
    "\n",
    "    if not os.path.exists(output_clade_dir):\n",
    "        os.makedirs(output_clade_dir, exist_ok=True)\n",
    "\n",
    "    output_fasta_path = f'{output_clade_dir}/{strain_clade}.filtered.fasta'\n",
    "\n",
    "    # filter fasta files\n",
    "\n",
    "    with open(strainphlan_fasta_path,'r') as input_fasta, open(output_fasta_path, 'w') as output_fasta:\n",
    "        for record in SeqIO.parse(input_fasta, 'fasta'):\n",
    "            if record.id in sample_list:\n",
    "                SeqIO.write(record, output_fasta, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ee6ca5b-0cc8-4c19-be25-1b7a07855846",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function for run RAxML\n",
    "\n",
    "def run_raxml(strain_clade, strain_out_dir=os.path.abspath(strain_ana)):\n",
    "    \"\"\"\n",
    "    run raxml on filtered fasta, THIS CAN ONLY BE RUN ON BIOBAKERY ENVIRONMENT ON MERGE\n",
    "    --------------------------------------------------\n",
    "    strain_clade: str\n",
    "        strain clade name\n",
    "    strain_out_dir: str\n",
    "        folder to store the clade folders for the filtered fasta\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # set up the command for raxml\n",
    "    raxml_cmd = (f'[YOUR CONDA ENVIRONMENT FOLDER]/bin/raxmlHPC-PTHREADS-SSE3 -p 1989 -m GTRCAT -T 8 '\n",
    "                 f'-w {strain_out_dir}/{strain_clade} '\n",
    "                 f'-s {strain_out_dir}/{strain_clade}/{strain_clade}.filtered.fasta '\n",
    "                 f'-n {strain_clade}.raxml.tre')\n",
    "    # split the command to small trunks of strings\n",
    "    raxml_cmd_list = raxml_cmd.split(' ')\n",
    "\n",
    "    # run the command\n",
    "    subprocess.run(raxml_cmd_list, shell=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "088004b5-fa66-43c2-bb92-c5c796c6146c",
   "metadata": {},
   "source": [
    "## RUN THE BLOCK BELOW IF WANNA USE 0.1 THRESHOLD FOR STRAINS NOT IN VALLES 2022 PAPER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "82a18ce9-7abc-4e6c-8669-bff41891637b",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp strain_transmission.py [YOUR CONDA ENVIRONMENT FOLDER]/lib/python3.9/site-packages/metaphlan/utils/strain_transmission.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "54d69d63-0454-4118-b1ed-dff4ad629944",
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_transmission(strain_clade,\n",
    "                        strain_out_dir=os.path.abspath(strain_ana),\n",
    "                        transmission_meta_file=f'{os.path.abspath(strain_ana)}/strain_transmission_metadata.tsv'):\n",
    "    \"\"\"\n",
    "    run transmission detection on the clade, output the return code\n",
    "    THIS CAN ONLY BE RUN ON BIOBAKERY ENVIRONMENT ON MERGE\n",
    "    --------------------------------------------------\n",
    "    strain_clade: str\n",
    "        strain clade name\n",
    "    strain_out_dir: str\n",
    "        folder to store the clade folders for the filtered fasta\n",
    "    transmission_meta_file: str\n",
    "        the metadata for transmission\n",
    "    \n",
    "    \"\"\"\n",
    "    # set up the command for transmission detection\n",
    "    # if b.longum subspecies, use t__SGB17248\n",
    "    if strain_clade in ['t__subsp.infantis', 't__subsp.longum']:\n",
    "        str_trans_cmd = (f'[YOUR CONDA ENVIRONMENT FOLDER]/bin/strain_transmission.py '\n",
    "                 f'-t {strain_out_dir}/{strain_clade}/RAxML_result.{strain_clade}.raxml.tre '\n",
    "                 f'-m {transmission_meta_file} '\n",
    "                 f'-o {strain_out_dir}/{strain_clade} '\n",
    "                 f'--sgb_id t__SGB17248')\n",
    "    else:\n",
    "        str_trans_cmd = (f'[YOUR CONDA ENVIRONMENT FOLDER]/bin/strain_transmission.py '\n",
    "                 f'-t {strain_out_dir}/{strain_clade}/RAxML_result.{strain_clade}.raxml.tre '\n",
    "                 f'-m {transmission_meta_file} '\n",
    "                 f'-o {strain_out_dir}/{strain_clade} '\n",
    "                 f'--sgb_id {strain_clade}')\n",
    "\n",
    "    str_trans_cmd_list = str_trans_cmd.split(' ')\n",
    "    run_td = subprocess.run(str_trans_cmd_list, shell=False, stderr=True)\n",
    "\n",
    "    return run_td.returncode"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
